server/src/main/java/org/apache/lucene/index/BaseSingleStarTreeBuilder.java

```diff
@@ -0,0 +1,660 @@
+        node.children = children;
+
+        // Construct star-node if required
+        if (!skipStarNodeCreationForDimensions.contains(childDimensionId) && children.size() > 1) {
+            children.put((long) StarTreeBuilderUtils.ALL, constructStarNode(startDocId, endDocId, childDimensionId));
+        }
+
+        // Further split on child nodes if required
+        for (StarTreeBuilderUtils.TreeNode child : children.values()) {
+            if (child.endDocId - child.startDocId > maxLeafDocuments) {
+                constructStarTree(child, child.startDocId, child.endDocId);
+            }
+        }
+    }
+
+    /**
+     * Constructs non star tree nodes
+     *
+     * @param startDocId  start document id
+     * @param endDocId    end document id
+     * @param dimensionId id of the dimension in the star tree
+     * @return root node with non-star nodes constructed
+     * @throws IOException throws an exception if we are unable to construct non-star nodes
+     */
+    private Map<Long, StarTreeBuilderUtils.TreeNode> constructNonStarNodes(int startDocId, int endDocId, int dimensionId)
+        throws IOException {
+        Map<Long, StarTreeBuilderUtils.TreeNode> nodes = new HashMap<>();
+        int nodeStartDocId = startDocId;
+        long nodeDimensionValue = getDimensionValue(startDocId, dimensionId);
+        for (int i = startDocId + 1; i < endDocId; i++) {
+            long dimensionValue = getDimensionValue(i, dimensionId);
+            if (dimensionValue != nodeDimensionValue) {
+                StarTreeBuilderUtils.TreeNode child = getNewNode();
+                child.dimensionId = dimensionId;
+                child.dimensionValue = nodeDimensionValue;
+                child.startDocId = nodeStartDocId;
+                child.endDocId = i;
+                nodes.put(nodeDimensionValue, child);
+
+                nodeStartDocId = i;
+                nodeDimensionValue = dimensionValue;
+            }
+        }
+        StarTreeBuilderUtils.TreeNode lastNode = getNewNode();
+        lastNode.dimensionId = dimensionId;
+        lastNode.dimensionValue = nodeDimensionValue;
+        lastNode.startDocId = nodeStartDocId;
+        lastNode.endDocId = endDocId;
+        nodes.put(nodeDimensionValue, lastNode);
+        return nodes;
+    }
+
+    /**
+     * Constructs star tree nodes
+     *
+     * @param startDocId  start document id
+     * @param endDocId    end document id
+     * @param dimensionId id of the dimension in the star tree
+     * @return root node with star nodes constructed
+     * @throws IOException throws an exception if we are unable to construct non-star nodes
+     */
+    private StarTreeBuilderUtils.TreeNode constructStarNode(int startDocId, int endDocId, int dimensionId) throws IOException {
+        StarTreeBuilderUtils.TreeNode starNode = getNewNode();
+        starNode.dimensionId = dimensionId;
+        starNode.dimensionValue = StarTreeBuilderUtils.ALL;
```

Comment:
We might need to check if instead of a constant, we can have a boolean value to denote whether its star node or not.

Reply:
Will be adding a boolean flag